databuilder/query_model.py

```diff
@@ -0,0 +1,388 @@
+#
+# The Query Model consists of operations on "frames" and "series". A frame is a table-like
+# thing that has rows and columns, and a series is a column-like thing that contains a
+# sequence of values. Frames can be created from SQL tables, and then various filtering,
+# transformation and sorting operations can be applied to produce new frames. Likewise,
+# a series can be created from a SQL column and then transformed and combined with
+# others to produce new series.
+#
+# Central to the Query Model design is that these frames and series have an additional
+# property called their "dimension": do they contain at most one row per patient, or
+# might they contain many rows?
+
+
+class Node:
+    "Abstract base class for all objects in the Query Model"
+
+    def __init_subclass__(cls, **kwargs):
+        # All nodes in the query model are frozen dataclasses
+        dataclasses.dataclass(cls, frozen=True)
+
+    def __post_init__(self):
+        # validate the things which have to be checked dynamically
+        validate_node(self)
+
+
+class Frame(Node):
+    ...
+
+
+class Series(Node):
+    def __class_getitem__(cls, type_):
+        # Series have an "inner" type denoting the type of value they contain e.g
+        # `Series[int]`. This is the method which makes that syntax work. At the moment
+        # we don't do anything with this type, we just return the original class
+        # unmodified. Later we will handle this properly and validate the types.
+        return cls
+
+
+class OneRowPerPatientFrame(Frame):
+    ...
+
+
+class ManyRowsPerPatientFrame(Frame):
+    ...
+
+
+class OneRowPerPatientSeries(Series):
+    ...
+
+
+class ManyRowsPerPatientSeries(Series):
+    ...
+
+
+class SortedFrame(ManyRowsPerPatientFrame):
+    ...
+
+
+# OPERATIONS
+#
+# Operations indicate the kind of thing they return by subclassing one of the basic
+# types above.
+
+
+# `Value` wraps a single static value in a one-row-per-patient series which just has
+# that value for every patient. This simplifies the signature of all other operations
+# which no longer need to care about static values.
+class Value(OneRowPerPatientSeries[T]):
+    value: T
+
+
+class SelectTable(ManyRowsPerPatientFrame):
+    name: str
+    schema: Mapping[str, Any] = DEFAULT_SCHEMA
+
+
+class SelectPatientTable(OneRowPerPatientFrame):
+    name: str
+    schema: Mapping[str, Any] = DEFAULT_SCHEMA
+
+
+class SelectColumn(Series):
+    source: Frame
+    name: str
+
+
+class Filter(ManyRowsPerPatientFrame):
+    source: ManyRowsPerPatientFrame
+    condition: Series[bool]
+
+
+class Sort(SortedFrame):
+    source: ManyRowsPerPatientFrame
+    sort_by: Series[Any]
+
+
+class PickOneRowPerPatient(OneRowPerPatientFrame):
+    source: SortedFrame
+    position: Position
+
+
+# An aggregation is any operation which accepts many-rows-per-patient series and returns
+# a one-row-per-patient series. Below are all available aggregations (using a class as a
+# namespace).
```

Comment:
Do we intend to add aggregations in the future? `Mean`, `Median`, and `Mode` spring to mind.

Reply:
Yes, absolutely. The same with functions too. I should have said "_currently_ available".